pkg/library/container/mountSources.go

```diff
@@ -84,20 +88,43 @@ func handleMountSources(k8sContainer *corev1.Container, devfileContainer *dw.Con
 		Name:  devfileConstants.ProjectsSourceEnvVar,
 		Value: path.Join(sourceMapping, projectsSourcePath),
 	})
+
+	return nil
 }
 
-// getProjectSourcePath gets the path, relative to PROJECTS_ROOT, that should be used for the PROJECT_SOURCE env var
-func getProjectSourcePath(projects []dw.Project) string {
-	projectPath := ""
+// getProjectSourcePath gets the path, relative to PROJECTS_ROOT, that should be used for the PROJECT_SOURCE env var.
+// Returns an error if there was a problem retrieving the selected starter project.
+//
+// The project source path is determined based on the following priorities:
+//
+//  1. If the workspace has at least one regular project, the first one will be selected.
+//     If the first project has a clone path, it will be used, otherwise the project's name will be used as the project source path.
+//
+//  2. If the workspace has a starter project that is selected, its name will be used as the project source path.
+//
+//  3. If the workspace has at least one starter project, but none are selected,
+//     the name of the first starter project will be used as the project source path.
+//
+//  4. Otherwise, the returned project source path will be an empty string.
+func getProjectSourcePath(workspace *dw.DevWorkspaceTemplateSpec) (string, error) {
+	projects := projectslib.GetAllProjects(workspace)
 	if len(projects) > 0 {
+		selectedStarterProject, err := projectslib.GetStarterProject(workspace)
+		if err != nil {
+			return "", err
+		}
+		// Only use the selected starterProject if we have no regular projects
+		if selectedStarterProject != nil && len(workspace.Projects) == 0 {
+			return selectedStarterProject.Name, nil
+		}
+
 		firstProject := projects[0]
 		if firstProject.ClonePath != "" {
-			projectPath = firstProject.ClonePath
-		} else {
-			projectPath = firstProject.Name
+			return firstProject.ClonePath, nil
 		}
+		return firstProject.Name, nil
 	}
-	return projectPath
+	return "", nil
```

Comment:
This still looks strange to me -- I think there's some duplication of work here since both `projectslib.GetAllProjects()` and `projectslib.GetStarterProject()` check for the selected starter project, so we're searching through `workspace.StarterProjects` twice. We're also aliasing `projects` to refer to "all projects", which is confusing given that there's a `workspace.Projects`

Would it make more sense to do something like

```go
// If there are any projects, return the first one's clone path
if len(workspace.Projects) > 0 {
  return projectslib.GetClonePath(workspace.Projects[0])
}
// No projects, check for starter project
starterProject, err := projectslib.GetStarterProject(workspace)
if err != nil {
  return "", err
} else if starterProject == nil {
  return "", nil
}
return starterProject.Name
``` 

This approach also won't throw an error here (but maybe will elsewhere) if there's a regular project but the selected starter project is misconfigured.

Reply:
I agree this is a better approach and is a lot more obvious as to what's going on (especially without requring the list of projects to be sorted with regular projects first, then starter projects second). This also eliminates the need for `projectslib.GetAllProjects()`, which felt a bit awkward of a function anyhow.